R/conditions.R

```diff
@@ -180,6 +189,38 @@ allow_lossy_cast <- function(expr, x_ptype = NULL, to_ptype = NULL) {
   )
 }
 
+maybe_warn_deprecated_lossy_cast <- function(x, to) {
+  # Returns `TRUE` if `allow_lossy_cast()` is on the stack and accepts
+  # to handle the condition
+  handled <- withRestarts(
+    vctrs_restart_error_cast_lossy = function() TRUE,
+    {
+      cnd_signal(cnd("vctrs_error_cast_lossy", x = x, to = to))
+      FALSE
+    }
+  )
+
+  if (!handled) {
+    from <- vec_ptype_abbr(x)
+    to <- vec_ptype_abbr(to)
+    warn_deprecated(paste_line(
+      glue::glue("We detected a lossy transformation from `{ from }` to `{ to }`."),
+      "The result will contain lower-resolution values or missing values.",
+      "To suppress this warning, wrap your code with `allow_lossy_cast()`:",
+      "",
+      "  # Allow all lossy transformations:",
+      "  vctrs::allow_lossy_cast(mycode())",
+      "",
+      "  # Allow only a specific transformation:",
+      "  vctrs::allow_lossy_cast(mycode(), x_ptype = from, to_ptype = to)",
```

Comment:
```suggestion
      "  vctrs::allow_lossy_cast(mycode(), x_ptype = from(), to_ptype = to())",
```

Not entirely sure about that, maybe skip entirely because it's hard to be accurate here. The warning is already long enough.

Reply:
You can just pass the full objects, we take the prototype automatically.

Reply:
I was wondering if we should rename `ptype` to `type` for that reason.

Reply:
I think `ptype` is fine; an object is by definition its own prototype.